File: packages/core/src/forces/natural.ts
/**
 * Natural primitives (§20.10) — real field laws, not the designed UI falloffs.
 *
 * The canonical nine (§6) are *designed*: finite range, soft `(1 − d/d_max)ⁿ`
 * falloff, tuned for legible motion on content bodies. These are *natural*: a true
 * softened inverse-square law in the sim unit system (§20.10). `gravity` and
 * `charge` are the **same kernel** — only the source scalar differs (mass ≥ 0 vs.
 * signed charge), which is the one unification the spec says is worth coding once.
 *
 * Opt-in: a body only feels them via `data-body="gravity"` / `data-body="charge"`,
 * so registering them changes nothing on a page that doesn't ask for them.
 */

import type { Body, Particle, Env, Force } from '../core/types.ts';
import type { Registry } from '../core/registry.ts';

/**
 * The shared softened inverse-square kernel (§20.10): `s / (d² + ε²)` along the
 * unit vector toward the body, then clamp speed to the unit system's `c` (the
 * hard velocity cap that *is* the speed of light in-sim).
 *
 * Plummer softening `ε = r_s = 2GM/c²` keeps the force finite at the core while
 * staying a true `1/d²` law far out — exactly how N-body sims avoid the singularity.
 * `s` is the signed source strength: `+GM` for gravity (always attractive),
 * `−σ·q·GM` for charge (like repels, opposite attracts).
 *
 * Sign note: this engine's `e.dx/e.dy` point from the particle *toward* the body,
 * so `+s` pulls inward (gravity attracts). Charge negates `σ·q·GM` so that like
 * signs push outward — the §20.3 formula written for an outward-pointing `û`.
 */
export function inverseSquare(b: Body, p: Particle, e: Env, s: number): void {
  if (e.dist >= b.range) return; // practical cutoff radius (an N-body softening too)
  const rs = (2 * e.G * b.M) / (e.c * e.c); // Schwarzschild radius → softening ε (§20.10)
  const f = s / (e.dist * e.dist + rs * rs); // s/(d²+ε²)
  p.vx += (e.dx / e.dist) * f;
  p.vy += (e.dy / e.dist) * f;
  clampToC(p, e.c);
}

/** Clamp a particle's speed to the unit system's `c` — the hard velocity cap that
 *  IS the in-sim speed of light (§20.10). Shared by the natural primitives. */
function clampToC(p: Particle, c: number): void {
  const sp = Math.hypot(p.vx, p.vy);
  if (sp > c) {
    p.vx = (p.vx / sp) * c;
    p.vy = (p.vy / sp) * c;
  }
}

/** §20.10 — true softened inverse-square: `F = GM·d̂/(d²+ε²)`, always attractive. */
export const gravity: Force = {
  token: 'gravity',
  label: 'Gravity',
  apply(b, p, e) {
    inverseSquare(b, p, e, e.G * b.M); // GM, mass-sourced (M ≥ 0 → pulls in)
  },
  meta: { desc: 'true softened inverse-square gravity (a real 1/d² law)' },
};

/** §20.3/§20.10 — the signed sibling of gravity; same kernel, sign sets direction. */
export const charge: Force = {
  token: 'charge',
  label: 'Charge',
  apply(b, p, e) {
    const q = p.charge ?? 0; // neutral matter ignores charge fields
    if (q === 0) return;
    // F = σ·q·GM/(d²+ε²); σ = body sign (data-spin), GM = G·M. Negated for the
    // inward-pointing kernel so like signs repel and opposite signs attract.
    inverseSquare(b, p, e, -(b.spin * q * e.G * b.M));
  },
  meta: { desc: 'signed inverse-square — like repels, opposite attracts' },
};

/**
 * §20.10 — the Lorentz force on a moving charge. In 2D the magnetic field `B` is a
 * scalar out of the plane, so the force is perpendicular to velocity: it curves a
 * particle's path into a circle (cyclotron radius `r_L = m|v|/(qB)`) **without doing
 * work** — speed is preserved, only the heading turns. The body's `spin` sets the
 * out-of-plane sense (which way it curls); `strength` is `|B|`. Acts only on charged,
 * *moving* matter; neutral particles pass straight through.
 */
export const magnetism: Force = {
  token: 'magnetism',
  label: 'Magnetism',
  apply(b, p, e) {
    if (e.dist >= b.range) return; // inside the field region
    const q = p.charge ?? 0;
    if (q === 0) return; // the Lorentz force needs charge
    // Exact rotation by θ = qB per frame — preserves |v| to floating-point precision.
    // The Euler form (p.vx += -vy*f; p.vy += vx*f) passes the ⟂ test but accumulates
    // speed as sqrt(1+(qB)²)^N, which grows noticeably at strength > 0.1.
    const theta = q * b.spin * b.strength;
    const cs = Math.cos(theta);
    const sn = Math.sin(theta);
    const vx0 = p.vx;
    p.vx = vx0 * cs - p.vy * sn;
    p.vy = vx0 * sn + p.vy * cs;
  },
  meta: { desc: 'Lorentz force — curves a moving charge perpendicular to its velocity' },
};

/**
 * The Langevin noise amplitude `σ = √(2·k_B·T·γ)` (§20.10). In sim units `k_B = γ = 1`,
 * so `σ = √(2T)`; negative `T` is floored to 0 (no imaginary kicks). Pure, so the
 * fluctuation–dissipation law itself is golden-tested apart from the RNG.
 */
export function thermalSigma(T: number): number {
  return Math.sqrt(2 * Math.max(0, T));
}

/**
 * §20.10 — `thermal`: Langevin/Brownian agitation, the *honest* `wander`. Each frame
 * a charge-free Gaussian kick `v += σ·ξ` (ξ ~ N(0,1) per axis) jiggles matter, with
 * `σ = √(2T)`. Paired with `drag` (`−γv`) it's a **thermostat** — fluctuation–
 * dissipation, the swarm equilibrates at temperature `T`. `T` is the body's strength,
 * eased to a localized hot spot by a `(1 − d/d_max)` falloff (sourceable from heat or
 * scroll energy). Box–Muller turns two uniforms into one isotropic 2-D kick.
 */
export const thermal: Force = {
  token: 'thermal',
  label: 'Thermal',
  apply(b, p, e) {
    if (e.dist >= b.range) return;
    const falloff = 1 - e.dist / b.range; // localized: hotter nearer the source
    const sigma = thermalSigma(b.strength * falloff);
    if (sigma === 0) return;
    // Box–Muller: (u1, u2) → one isotropic N(0,1) pair, scaled by σ.
    const u1 = Math.random() || 1e-9; // avoid log(0)
    const mag = sigma * Math.sqrt(-2 * Math.log(u1));
    const ang = 2 * Math.PI * Math.random();
    p.vx += mag * Math.cos(ang);
    p.vy += mag * Math.sin(ang);
    if (b.on) p.heat = Math.max(p.heat, falloff * 0.4);
    clampToC(p, e.c);
  },
  meta: { desc: 'Langevin/Brownian agitation — a real temperature in the medium' },
};

/**
 * §20.10 — `collide`: elastic pairwise collision (granular / billiard), the hard-sphere
 * complement to the smooth, box-bound `wall`. For each neighbour `q` whose disc
 * overlaps `p`'s (`d < r_p + r_q`, radius ≈ `size`) and that is *approaching*, the pair
 * exchanges normal momentum: `p` takes a half-impulse along the contact normal `n`,
 * `q` the other half on its own turn — so momentum is conserved and, at `e = 1`, energy
 * too. `strength` is the restitution `e ∈ [0,1]`. Class [B] — uses `env.neighbors`.
 */
export const collide: Force = {
  token: 'collide',
  label: 'Collide',
  apply(b, p, e) {
    if (e.dist >= b.range) return; // collisions resolve within the body's region
    const restitution = Math.max(0, Math.min(1, b.strength));
    const pr = Math.max(1, p.size);
    for (const q of e.neighbors(p, pr * 4)) {
      const qr = Math.max(1, q.size);
      const nx = p.x - q.x;
      const ny = p.y - q.y;
      const d = Math.hypot(nx, ny);
      if (d >= pr + qr || d < 1e-6) continue; // not in contact
      const ux = nx / d;
      const uy = ny / d;
      const relN = (p.vx - q.vx) * ux + (p.vy - q.vy) * uy;
      if (relN >= 0) continue; // separating already → no impulse
      // resolve the pair symmetrically in one pass (equal & opposite impulses) so the
      // result is momentum-conserving and order-independent — the integrator processes
      // particles sequentially, so applying only to `p` and trusting `q`'s later turn
      // double-counts (q would read p's already-changed velocity). After this the pair
      // is separating, so q's own apply skips it.
      const j = (1 + restitution) * 0.5 * relN;
      p.vx -= j * ux;
      p.vy -= j * uy;
      q.vx += j * ux;
      q.vy += j * uy;
    }
  },
  meta: { desc: 'elastic pairwise collision — the hard-sphere billiard force' },
};

/**
 * §20.10 — `diffuse` (class [C], over the scalar `grid`): the pheromone/stigmergy
 * field. Each frame a particle lays a mark into the shared `diffuse` grid (which the
 * engine blurs via `∂φ/∂t = D∇²φ`) and steers *up* the local gradient, following the
 * smeared trail toward where matter has gathered. `strength` sets both the deposit and
 * the follow gain. Self-organizing trails emerge from the deposit↔blur↔follow loop.
 */
export const diffuse: Force = {
  token: 'diffuse',
  label: 'Diffuse',
  apply(b, p, e) {
    if (e.dist >= b.range) return;
    const g = e.grid('diffuse');
    g.deposit(p.x, p.y, b.strength); // lay a mark
    const grad = g.gradient(p.x, p.y); // follow the blurred trail up-gradient
    p.vx += grad.x * b.strength;
    p.vy += grad.y * b.strength;
  },
  meta: { desc: 'pheromone field — deposit a mark and follow the diffused gradient' },
};

/** Frames between emitted shocks while a propagate body is engaged. A *pulse train* — not a
 *  continuous drip — is what keeps it a travelling wave: between pulses the grid radiates and
 *  damps, so no standing bump builds at the source (that bump is what used to pull matter IN). */
const WAVE_PULSE_PERIOD = 12;
/** How hard a passing wavefront carries matter outward (radiation pressure gain). */
const WAVE_PUSH = 7;

/**
 * §20.10 — `propagate` (class [C], over a wave-mode `grid`): a travelling disturbance,
 * `∂²φ/∂t² = c²∇²φ`. An engaged body injects an impulsive shock at its centre (via the
 * body-level `source` hook, once per frame — not once per particle), and the grid carries it
 * outward as a real expanding ring. Matter **rides the front out**: where the wavefront is
 * passing (`|∇φ|` is steep) a particle is pushed radially *away from the source*, so the
 * expanding shock sweeps matter outward with it — radiation pressure, not an inward pull.
 *
 * Why a pulse train, not a continuous deposit: depositing every frame builds a standing φ bump
 * at the centre, whose gradient points inward — `v += ∇φ` then sucks matter toward the source
 * (the opposite of a wave). Emitting a shock every `WAVE_PULSE_PERIOD` frames lets each ring
 * radiate and damp away, so the field stays a sequence of outgoing fronts.
 */
export const propagate: Force = {
  token: 'propagate',
  label: 'Propagate',
  source(b, e) {
    if (!b.on) return; // only an engaged body emits
    if (e.frameN % WAVE_PULSE_PERIOD !== 0) return; // a shock train, once per period (body-level)
    e.grid('wave-propagate').deposit(b.cx, b.cy, b.strength); // 'wave…' name → wave stepping
  },
  apply(b, p, e) {
    if (e.dist >= b.range) return;
    const g = e.grid('wave-propagate');
    const grad = g.gradient(p.x, p.y);
    const act = Math.hypot(grad.x, grad.y); // wavefront activity — steep where a front is passing
    if (act < 1e-6) return; // no front here → coast (the wave has moved on or not yet arrived)
    // ride the front: pushed radially OUTWARD (e.dx/e.dy point toward the body, so negate).
    const ux = -e.dx / e.dist;
    const uy = -e.dy / e.dist;
    p.vx += ux * act * b.strength * WAVE_PUSH;
    p.vy += uy * act * b.strength * WAVE_PUSH;
    clampToC(p, e.c);
  },
  meta: { desc: 'a travelling wave — a shock train expands from the source, sweeping matter out' },
};

/**
 * Memory (class [C], over a slow-decaying `memory` grid) — the field remembers.
 * Each frame a particle lays occupancy where it sits, into a grid that barely blurs
 * and fades slowly; the body's pull is then amplified by how worn the spot is. So
 * `M(x) += λ` and the effective force `×= (1 + μ·M)`: frequently-travelled routes
 * deepen and pull harder, and channels wear in over time.
 */
export const memory: Force = {
  token: 'memory',
  label: 'Memory',
  apply(b, p, e) {
    if (e.dist >= b.range) return;
    const g = e.grid('memory'); // 'memory' name → slow-decay stepping
    g.deposit(p.x, p.y, b.strength * 0.15); // wear the path where matter sits
    const amp = 1 + 0.5 * g.sample(p.x, p.y); // worn paths pull harder (1 + μ·M)
    const f = (1 - e.dist / b.range) ** 2 * b.strength * 0.5 * amp;
    p.vx += (e.dx / e.dist) * f;
    p.vy += (e.dy / e.dist) * f;
  },
  meta: { desc: 'the field remembers — occupancy wears in paths that pull harder' },
};

/** The natural primitives, in spec order (§20.10). */
export const naturalForces: readonly Force[] = [
  gravity,
  charge,
  magnetism,
  thermal,
  collide,
  diffuse,
  propagate,
  memory,
];

/** Register the natural primitives on a registry (§4) — opt-in, alongside the nine. */
export function registerNaturalForces(reg: Registry): void {
  for (const f of naturalForces) reg.force(f);
}

File: rust/crates/fundamental-core/src/forces/natural.rs
//! Natural primitives (§20.10) — real field laws, mirrors `packages/core/src/forces/natural.ts`.
//!
//! Where the canonical nine are *designed* (finite range, soft `(1 − d/d_max)ⁿ` falloff), these are
//! *natural*: a true softened inverse-square law in the sim unit system. `gravity` and `charge` are the
//! **same kernel** — only the source scalar differs (mass ≥ 0 vs. signed charge). Opt-in: a body only
//! feels them via its tokens, so registering them changes nothing on a body that doesn't ask.
//!
//! Ported here: the four that read only the per-particle `env` (`gravity`, `charge`, `magnetism`,
//! `thermal`). Deferred with their subsystems: `collide` (neighbour query), `diffuse` / `propagate` /
//! `memory` (scalar grid). The renderable `field()` structure hooks (dipole/monopole) land with the
//! field-line/streamline layer.

use crate::engine::{Body, Env, Force, Particle};
use std::f64::consts::PI;

/// Clamp a particle's speed to the unit system's `c` — the hard velocity cap that IS the in-sim speed
/// of light (§20.10). Shared by the natural primitives.
fn clamp_to_c(p: &mut Particle, c: f64) {
    let sp = p.velocity.length();
    if sp > c {
        p.velocity = p.velocity * (c / sp);
    }
}

/// The shared softened inverse-square kernel (§20.10): `s / (d² + ε²)` along the unit vector toward the
/// body, then clamp to `c`. Plummer softening `ε = r_s = 2GM/c²` keeps the force finite at the core
/// while staying a true `1/d²` law far out. `s` is the signed source strength.
fn inverse_square(b: &Body, p: &mut Particle, e: &Env, s: f64) {
    if e.dist >= b.range {
        return; // practical cutoff radius
    }
    let rs = (2.0 * e.g * b.source_mass) / (e.c * e.c); // Schwarzschild radius → softening ε
    let f = s / (e.dist * e.dist + rs * rs);
    p.velocity.x += (e.vector.x / e.dist) * f;
    p.velocity.y += (e.vector.y / e.dist) * f;
    if e.vector.z != 0.0 {
        p.velocity.z += (e.vector.z / e.dist) * f;
    }
    clamp_to_c(p, e.c);
}

/// The Langevin noise amplitude `σ = √(2·k_B·T·γ)`; in sim units `k_B = γ = 1`, so `σ = √(2T)`.
pub fn thermal_sigma(t: f64) -> f64 {
    (2.0 * t.max(0.0)).sqrt()
}

/// §20.10 — true softened inverse-square: `F = GM·d̂/(d²+ε²)`, always attractive.
pub struct Gravity;

impl Force for Gravity {
    fn token(&self) -> &'static str {
        "gravity"
    }
    fn label(&self) -> &'static str {
        "Gravity"
    }
    fn apply(&self, b: &Body, p: &mut Particle, e: &mut Env) {
        inverse_square(b, p, e, e.g * b.source_mass); // GM, mass-sourced (M ≥ 0 → pulls in)
    }
}

/// §20.3/§20.10 — the signed sibling of gravity; same kernel, sign sets direction (like repels,
/// opposite attracts). Acts only on charged matter.
pub struct Charge;

impl Force for Charge {
    fn token(&self) -> &'static str {
        "charge"
    }
    fn label(&self) -> &'static str {
        "Charge"
    }
    fn apply(&self, b: &Body, p: &mut Particle, e: &mut Env) {
        let q = p.charge;
        if q == 0.0 {
            return; // neutral matter ignores charge fields
        }
        // F = σ·q·GM/(d²+ε²); negated for the inward-pointing kernel so like signs repel.
        inverse_square(b, p, e, -(b.spin * q * e.g * b.source_mass));
    }
}

/// §20.10 — the Lorentz force on a moving charge: a rotation perpendicular to velocity (a cyclotron
/// curl) that preserves speed exactly. Acts only on charged, *moving* matter; the body's `spin` sets
/// the out-of-plane sense, `strength` is `|B|`, graded by a `(1 − d/r)` falloff.
pub struct Magnetism;

impl Force for Magnetism {
    fn token(&self) -> &'static str {
        "magnetism"
    }
    fn label(&self) -> &'static str {
        "Magnetism"
    }
    fn apply(&self, b: &Body, p: &mut Particle, e: &mut Env) {
        if e.dist >= b.range {
            return;
        }
        let q = p.charge;
        if q == 0.0 {
            return; // the Lorentz force needs charge
        }
        // exact rotation by θ = q·spin·B·falloff — preserves |v| to floating-point precision.
        let falloff = 1.0 - e.dist / b.range;
        let theta = q * b.spin * b.strength * falloff;
        let (sn, cs) = theta.sin_cos();
        let vx0 = p.velocity.x;
        p.velocity.x = vx0 * cs - p.velocity.y * sn;
        p.velocity.y = vx0 * sn + p.velocity.y * cs;
    }
}

/// §20.10 — `thermal`: Langevin/Brownian agitation, the *honest* wander. A Gaussian kick `v += σ·ξ`
/// with `σ = √(2T)`, localized by a `(1 − d/d_max)` falloff. Box–Muller turns two uniforms (from the
/// seeded rng) into one isotropic kick, so a seeded thermal run is reproducible.
pub struct Thermal;

impl Force for Thermal {
    fn token(&self) -> &'static str {
        "thermal"
    }
    fn label(&self) -> &'static str {
        "Thermal"
    }
    fn apply(&self, b: &Body, p: &mut Particle, e: &mut Env) {
        if e.dist >= b.range {
            return;
        }
        let falloff = 1.0 - e.dist / b.range;
        let sigma = thermal_sigma(b.strength * falloff);
        if sigma == 0.0 {
            return;
        }
        // Box–Muller from the injected rng — drawn in the same order as JS for cross-plane parity.
        let u1 = {
            let u = e.rng();
            if u == 0.0 {
                1e-9
            } else {
                u
            }
        }; // avoid log(0)
        let mag = sigma * (-2.0 * u1.ln()).sqrt();
        let ang = 2.0 * PI * e.rng();
        p.velocity.x += mag * ang.cos();
        p.velocity.y += mag * ang.sin();
        // in a depth > 0 volume the kick is isotropic in 3D — a third gaussian leg (z-axis.md).
        if e.volume.z > 0.0 {
            let u2 = {
                let u = e.rng();
                if u == 0.0 {
                    1e-9
                } else {
                    u
                }
            };
            p.velocity.z += sigma * (-2.0 * u2.ln()).sqrt() * (2.0 * PI * e.rng()).cos();
        }
        if b.engaged {
            p.heat = p.heat.max(falloff * 0.4);
        }
        clamp_to_c(p, e.c);
    }
}
